### PR TITLE
develop to master

### PR DESCRIPTION
**Pull request by budgetpreneur**

### CODE REVIEW

File: src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void Signal(string channel, object message, Dictionary<string, object> metaData, Dictionary<string, object> externalQueryParam, PNCallback<PNPublishResult> callback)
+        {
+            if (string.IsNullOrEmpty(channel) || string.IsNullOrEmpty(channel.Trim()) || message == null)
+            {
+                throw new ArgumentException("Missing Channel or Message");
+            }
+
+            if (string.IsNullOrEmpty(config.PublishKey) || string.IsNullOrEmpty(config.PublishKey.Trim()) || config.PublishKey.Length <= 0)
```
Comment (davidnub) [inline src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs]: Should we also check for subscribe key?

File: src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            {
+                throw new MissingMemberException("Invalid publish key");
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentException("Missing userCallback");
```
Comment (davidnub) [inline src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs]: I thought we can allow empty callback and just pass `0` if one was not specified?

File: src/Api/PubnubApi/Model/Server/SubscribeMessage.cs [hunk 3]
```diff
@@ -28,6 +28,11 @@ internal class SubscribeMessage
         /// </summary>
         private object d { get; set; } = "";
 
+        /// <summary>
+        /// signal indicator
```
Comment (davidnub) [inline src/Api/PubnubApi/Model/Server/SubscribeMessage.cs]: This is actually a message type indicator

File: src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs (hunk 1 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs]: added check for SubscribeKey also. will merge to PR in few mins.

File: src/Api/PubnubApi/Model/Server/SubscribeMessage.cs (hunk 3 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/Model/Server/SubscribeMessage.cs]: corrected "signal" to "message type"

File: src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs (hunk 2 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs]: C# sdk never sent 0 for empty/null callback for any feature. If needed, we can review for future release.

**Pull request merged**